File: TypeScript/observer/subject.ts
import Observer from "./observer";

export default abstract class Subject {

    private observers: Observer[] = new Array<Observer>();

    public attach(observer: Observer) {
        var index = this.observers.indexOf(observer);
        if (index < 0) {
            this.observers.push(observer);
        }
    }

    public detach(observer: Observer) {
        var index = this.observers.indexOf(observer);
        if (index >= 0) {
            this.observers.splice(index, 1);
        }
    }

    public notify() {
        this.observers.forEach((o) => o.update(this));
    }
}

File: CSharp/Behavioral/Observer/Subject.cs
using System.Collections.Generic;

namespace Behavioral.Observer
{
    public abstract class Subject
    {
        private List<Observer> _observers = new List<Observer>();

        public virtual void Attach(Observer o)
        {
            _observers.Add(o);
        }

        public virtual void Detach(Observer o)
        {
            _observers.Remove(o);
        }

        public virtual void Notify()
        {
            foreach (var observer in _observers)
            {
                observer.Update(this);
            }
        }
    }
}